core/src/main/java/org/apache/calcite/sql/fun/SqlLambdaOperator.java

```diff
@@ -0,0 +1,115 @@
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlLambda;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.SqlOperatorBinding;
+import org.apache.calcite.sql.SqlSyntax;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.InferTypes;
+import org.apache.calcite.sql.type.OperandTypes;
+import org.apache.calcite.sql.type.ReturnTypes;
+import org.apache.calcite.sql.type.SqlOperandCountRanges;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.sql.validate.SqlValidatorScope;
+
+public class SqlLambdaOperator extends SqlOperator {
+
+  public static final SqlLambdaOperator INSTANCE = new SqlLambdaOperator();
+
+  //~ Instance fields --------------------------------------------------------
+
+  //~ Constructors -----------------------------------------------------------
+
+  /**
+   * Creates a SqlLambdaOperator.
+   */
+  SqlLambdaOperator() {
+    super(
+        "->",
+        SqlKind.LAMBDA,
+        32,
+        false,
+        ReturnTypes.ARG0_NULLABLE,
+        InferTypes.RETURN_TYPE,
+        OperandTypes.VARIADIC);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  public SqlOperandCountRange getOperandCountRange() {
+    return SqlOperandCountRanges.any();
+  }
+
+  @Override public SqlSyntax getSyntax() {
+    return SqlSyntax.SPECIAL;
+  }
+
+  public boolean checkOperandTypes(
+      SqlCallBinding callBinding,
+      boolean throwOnFailure) {
+    return true;
+  }
+
+  public void unparse(
+      SqlWriter writer,
+      SqlCall call,
+      int leftPrec,
+      int rightPrec) {
+    final SqlWriter.Frame frame = writer.startList("", "");
+    call.operand(0).unparse(writer, getLeftPrec(), getRightPrec());
+    writer.sep(getName());
+
+    call.operand(1).unparse(writer, getLeftPrec(), getRightPrec());
+    writer.endList(frame);
+  }
+
+  @Override public void validateCall(SqlCall call, SqlValidator validator,
+      SqlValidatorScope scope,
+      SqlValidatorScope operandScope) {
+    validateOperands(validator, scope, call);
+  }
+
+  public SqlCall createCall(
+      SqlLiteral functionQualifier,
+      SqlParserPos pos,
+      SqlNode... operands) {
+    assert functionQualifier == null;
+    assert operands.length == 2;
+    return new SqlLambda(pos, (SqlNodeList) operands[0], operands[1]);
+  }
+
+  @Override public RelDataType deriveType(
+      SqlValidator validator, SqlValidatorScope scope, SqlCall call) {
+    return validateOperands(validator, scope, call);
```

Comment:
Should implement the `checkOperandTypes` correctly to deduce the type, you can take `SqlCaseOperator` for a reference.

Reply:
SqlCaseOperator has different implementation Lambda type inference is derived from  LambdaOperandTypeChecker.java